lib/LANraragi/Plugin/Metadata/Hitomi.pm

```diff
@@ -0,0 +1,253 @@
+
+        # Matching URL Scheme like 'https://' is only for backward compatible purpose.
+        $galleryID = $1;
+        $logger->debug("Skipping search and using gallery $galleryID from source tag");
+    } else {
+
+        #Get Gallery ID by hand if the user didn't specify a URL
+        $galleryID = get_gallery_id_from_title( $lrr_info->{archive_title} );
+    }
+
+    # Did we detect a Hitomi gallery?
+    if ( defined $galleryID ) {
+        $logger->debug("Detected Hitomi gallery id is $galleryID");
+    } else {
+        $logger->info("No matching Hitomi Gallery Found!");
+        return ( error => "No matching Hitomi Gallery Found!" );
+    }
+
+    #If no tokens were found, return a hash containing an error message.
+    #LRR will display that error to the client.
+    if ( $galleryID eq "" ) {
+        $logger->info("No matching Hitomi Gallery Found!");
+        return ( error => "No matching Hitomi Gallery Found!" );
+    }
+
+    my %hashdata = get_tags_from_Hitomi( $galleryID, $gendertag, $savetitle);
+
+    $logger->info( "Sending the following tags to LRR: " . $hashdata{tags} );
+
+    #Return a hash containing the new metadata - it will be integrated in LRR.
+    return %hashdata;
+}
+
+######
+## Hitomi Specific Methods
+######
+
+sub get_gallery_id_from_title {
+
+    my ($title) = @_;
+
+    my $logger = get_plugin_logger();
+
+	$logger->debug("Attempting to parse id from title $title");
+    if ( $title =~ /\{?(\d+)\}?/gm ) {
+        $logger->debug("Got $1 from file.");
+        return $1;
+    }
+
+    return;
+}
+
+# retrieves js from Hitomi
+sub get_js_from_hitomi {
+
+    my ($gID) = @_;
+	
+	my $logger = get_plugin_logger();
+    
+    my $gJS = "https://ltn.hitomi.la/galleries/$gID.js";
+	
+	$logger->debug("Hitomi JS: $gJS");
+
+	my $ua = Mojo::UserAgent->new;
+	my $res = $ua->get($gJS)->result;
+	$logger->debug("Hitomi raw JS: ". $res->body);
+	
+	if ( $res->is_error ) {
+        return;
+    }
+	
+	my $jsonstring = "{}";
+	if ( $res->body =~ /var.*galleryinfo.*= (.*)/gmi ) {
+        $jsonstring = $1;
+    }
+
+    $logger->debug("Tentative new JSON: $jsonstring");
+
+	$logger->debug("Beginning JSON decode");
+    my $json = decode_json $jsonstring;
+	$logger->debug("JSON decode successful");
+    return $json;
+	
+}
+
+#Extract tags from Hitomi JSON
+sub get_tags_from_taglist {
+
+    my ($json, $gendertag) = @_;
+	
+	my $logger = get_plugin_logger();
+	
+	my @tags = ();
+
+	if(defined $json->{"tags"}) {
+		$logger->debug("Extracting tags array");
+		my @tags_list = @{ $json->{"tags"} };
+		
+		$logger->debug("Cycling tags array");
+		foreach my $tag (@tags_list) {
+			my $name = $tag->{"tag"};
+			my $male = $tag->{"male"};
+			my $female = $tag->{"female"};
+			
+			if($gendertag) {
+				if($male eq 1){
+					$name = $name . " ♂️"
+				}
+				
+				if($female eq 1){
+					$name = $name . " ♀️"
+				}
+			}
+
+			push( @tags, $name );
+		}
+	}
+	
+	if(defined $json->{"parodys"}) {
```

Comment:
I think you could indeed simplify the method a bit here: There's a fair bit of copy-pasted code that does the same thing except on a different array and for a different namespace.  

Ideally you should have a separate sub that'd look a bit like this (napkin perl, sorry if the syntax is incorrect at times!):  
```
sub extract_tags{
       my ( $sourcejson, $arrayname, $namespace) = @_;
       my @tags;

      $logger->debug("Extracting $arrayname array");
      my @list = @{ $json->{$arrayname} };
      $logger->debug("Cycling $arrayname array");

      foreach my $tag (@list) {
		my $name = $tag->{$namespace};
		push( @tags, "$namespace:$name");
      }
      return @tags;
}
```  

Which you could then call like so:  
```
push (@tags, extract_tags($json, "parodys", "parody");
push (@tags, extract_tags($json, "artists", "artist");
push (@tags, extract_tags($json, "groups", "group");
push (@tags, extract_tags($json, "characters", "character");
```